hasjob/models/filterset.py

```diff
@@ -0,0 +1,159 @@
+        from hasjob.views.helper import location_geodata
+
+        location_names = []
+        if translate_geonameids and self.location_geonameids:
+            location_dict = location_geodata(self.location_geonameids)
+            for geonameid in self.location_geonameids:
+                location_names.append(location_dict[geonameid]['name'])
+
+        return {
+            't': [jobtype.name for jobtype in self.types],
+            'c': [jobcategory.name for jobcategory in self.categories],
+            'l': location_names if translate_geonameids else self.location_geonameids,
+            'currency': self.pay_currency,
+            'pay': self.pay_cash_min,
+            'equity': self.equity,
+            'anywhere': self.remote_location
+        }
+
+    @classmethod
+    def from_filters(cls, board, filters):
+        basequery = cls.query.filter(cls.board == board)
+
+        if filters.get('t'):
+            basequery = basequery.join(
+                filterset_jobtype_table).join(
+                JobType).filter(JobType.name.in_(filters['t']))
+        else:
+            basequery = basequery.filter(~cls.id.in_(db.select([filterset_jobtype_table.c.filterset_id])))
+
+        if filters.get('c'):
+            basequery = basequery.join(
+                filterset_jobcategory_table).join(
+                JobCategory).filter(JobCategory.name.in_(filters['c']))
+        else:
+            basequery = basequery.filter(~cls.id.in_(db.select([filterset_jobcategory_table.c.filterset_id])))
+
+        if filters.get('l'):
+            basequery = basequery.filter(cls.location_geonameids == sorted(filters['l']))
+        else:
+            basequery = basequery.filter(cls.location_geonameids == None)
+
+        if filters.get('equity'):
+            basequery = basequery.filter(cls.equity == True)
+        else:
+            basequery = basequery.filter(cls.equity == False)
+
+        if filters.get('pay') and filters.get('currency'):
+            basequery = basequery.filter(cls.pay_cash_min == filters['pay'],
+                cls.pay_currency == filters['currency'])
+        else:
+            basequery = basequery.filter(cls.pay_cash_min == None, cls.pay_currency == None)
+
+        if filters.get('keywords'):
+            basequery = basequery.filter(cls.keywords == filters['keywords'])
+        else:
+            basequery = basequery.filter(cls.keywords == '')
+
+        if filters.get('anywhere'):
+            basequery = basequery.filter(cls.remote_location == True)
+        else:
+            basequery = basequery.filter(cls.remote_location == False)
+
+        return basequery.one_or_none()
+
+
+@event.listens_for(FilterSet, 'before_update')
+@event.listens_for(FilterSet, 'before_insert')
+def _format_and_validate(mapper, connection, target):
+    if target.location_geonameids:
+        target.location_geonameids = sorted(target.location_geonameids)
+
+    if FilterSet.from_filters(target.board, target.to_filters()):
```

Comment:
Is this safe? You'll querying the database in the middle of a flush.

Reply:
I didn't encounter problems with it; I'm not sure what the alternative is.